### PR TITLE
Reintroduce the Defaults page and the Reset buttons

### LINKED ISSUE #10430: Add Base layer back into the SUI as "Defaults"
# Description of the new feature/enhancement
We'd like to add Base layer back into the settings UI along with an [Extensions page](https://github.com/microsoft/terminal/issues/10436) to help avoid the conflicts between `profiles.defaults` and JSON fragment extensions.

The Extensions page will take some time to design, but we'd like to get started with getting Base layer back into the UI, so we'll track the progress on the Extensions page in another issue.

For now, let's add Base layer back into the settings UI with the following improvements:
- Change the name Base layer to Defaults
- Add the Defaults page back into the settings UI
- Add the reset buttons back into the profile pages
- Collect data on which settings are applied using Defaults 
- Collect data on how often the profile page reset buttons are used

Related: #9539 

# Proposed technical implementation details (optional)

### PR DESCRIPTION
**Pull request by DHowett**
This pull request brings back the "Base Layer" page, now renamed to
"Defaults", and the "Reset to inherited value" buttons. The scope of
inheritance for which buttons will display has been widened.

The button will be visible in the following cases:

The user has set a setting for the current profile, and it overrides...

1. ... something in profiles.defaults.
2. ... something in a Fragment Extension profile.
3. ... something from a Dynamic Profile Generator.
4. ... something from the compiled-in defaults.

Compared to the original implementation of reset arrows, cases (1), (3)
and (4) are new. Rationale:

(1) The user can see a setting on the Defaults page, and they need a way
    to reset back to it.

(3) Dynamic profiles are not meaningfully different from fragments, and
    users may need a way to reset back to the default value generated
    for WSL or PowerShell.

(4) The user can see a setting on the Defaults page, **BUT** they are
    not the one who created it. They *still* need a way to get back to
    it.

To support this, I've introduced another origin tag, "User", and renamed
"Custom" to "None". Due to the way origin/override detection works¹, we
cannot otherwise disambiguate between settings that came from the user
and settings that came from the compiled-in defaults.

Changes were required in TerminalSettings such that we could construct a
settings object with a profile that does not have a GUID. In making this
change, I fixed a bit of silliness where we took a profile, extracted
its guid, and used that guid to look up the same profile object. Oops.

I also fixed the PropertyChanged notifier to include the
XxxOverrideSource property.

The presence of the page and the reset arrows is restricted to
Preview- or Dev-branded builds. Stable builds will retain their current
behavior.

¹ `XxxOverrideSource` returns the profile *above* the current profile
  that holds a value for setting `Xxx`. When the value is the
  compiled-in value, `XxxOverrideSource` will be `null`. Since it's
  supposed to be the profile above the current profile, it will also be
  `null` if the profile contains a setting at this layer.
  In short, `null` means "user specified" *or* "compiled in". Oops.

Fixes #10430

Validation
----------

* [x] Tested Release build to make sure it's mostly arrow-free (apart from fragments)

### CODE REVIEW

File: src/cascadia/TerminalSettingsEditor/MainPage.xaml [hunk 1]
```diff
@@ -88,6 +88,15 @@
 
             <muxc:NavigationViewItemHeader x:Uid="Nav_Profiles" />
 
+            <muxc:NavigationViewItem x:Name="BaseLayerMenuItem"
+                                     x:Uid="Nav_ProfileDefaults"
+                                     Tag="GlobalProfile_Nav"
+                                     Visibility="{x:Bind ShowBaseLayerMenuItem}">
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsEditor/MainPage.xaml:94]: **IF** possible, bind `x:Load` instead.

File: src/cascadia/TerminalSettingsEditor/SettingContainer.cpp [hunk 2]
```diff
@@ -57,9 +57,9 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
             _SettingOverrideSourceProperty =
                 DependencyProperty::Register(
                     L"SettingOverrideSource",
-                    xaml_typename<bool>(),
+                    xaml_typename<IInspectable>(),
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsEditor/SettingContainer.cpp:60]: caught an oops! Source is not a bool :P

File: src/cascadia/TerminalSettingsEditor/SettingContainer.cpp [hunk 3]
```diff
@@ -182,35 +174,43 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
...
-            // base layer
-            // TODO GH#3818: When we add profile inheritance as a setting,
-            //               we'll need an extra conditional check to see if this
-            //               is the base layer or some other profile
-
-            // GH#9539: Base Layer has been removed from the Settings UI.
-            //          In the event that the Base Layer comes back,
-            //          return RS_(L"SettingContainer_OverrideMessageBaseLayer") instead
-            return {};
+            // from a fragment extension or generated profile
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsEditor/SettingContainer.cpp:211]: only `from a fragment extension`.

File: src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp [hunk 4]
```diff
@@ -893,6 +896,12 @@ void CascadiaSettings::_LayerOrCreateProfile(const Json::Value& profileJson)
             _allProfiles.Append(*profile);
         }
     }
+
+    if (_userDefaultProfileSettings)
+    {
+        // If we've loaded defaults{} we're in the "user settings" phase for sure
+        profile->Origin(OriginTag::User);
+    }
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:904]: @carlos-zamora this is the part that I am iffy on. I don't like it, but it is the easiest way to know that we are currently processing the user profile list compartment :|

Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:904]: This crashes when profile is null -- why is profile null? i guess it's.. a first-run scenario? I'm not actually sure...

Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:904]: ah, it's a profile that my DPG no longer emits

File: src/cascadia/TerminalSettingsEditor/MainPage.xaml (hunk 1 above)
Comment (DHowett) [reply src/cascadia/TerminalSettingsEditor/MainPage.xaml:94]: `x:Load` is not possible because of the dreaded "UnloadObject doesn't exist for some reason" error.

File: src/cascadia/TerminalSettingsEditor/SettingContainer.cpp [hunk 5]
```diff
@@ -182,35 +174,43 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
...
-            return {};
+            source = profile.Source();
+            originTag = profile.Origin();
         }
-        else if (originTag == Model::OriginTag::Generated)
+        else if (const auto& appearanceConfig{ settingOrigin.try_as<Model::AppearanceConfig>() })
         {
-            // from a dynamic profile generator
-            return {};
+            auto profile = appearanceConfig.SourceProfile();
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsEditor/SettingContainer.cpp:190]:
```suggestion
            const auto& profile = appearanceConfig.SourceProfile();
```

**Review by carlos-zamora — APPROVED** [review]
Looks good! Thanks for doing this!

File: src/cascadia/TerminalSettingsEditor/ViewModelHelpers.h [hunk 6]
```diff
@@ -39,33 +39,33 @@ struct ViewModelHelper
...
+#define _BASE_OBSERVABLE_PROJECTED_SETTING(target, name)                   \
+public:                                                                    \
+    auto name() const noexcept { return target.name(); };                  \
+    template<typename T>                                                   \
+    void name(const T& value)                                              \
+    {                                                                      \
+        if (name() != value)                                               \
+        {                                                                  \
+            target.name(value);                                            \
+            _NotifyChanges(L"Has" #name, L#name, L#name "OverrideSource"); \
```
[the remark this answers is not in the document]
Comment (DHowett) [reply src/cascadia/TerminalSettingsEditor/ViewModelHelpers.h:51]: XxxOverrideSource *is* our parent. It changes (goes from returning null to non-null) if the value of HasXxx changes. Technically it doesn't matter if we notify on it because the listener is also listening for HasXxx changing.

File: src/cascadia/TerminalSettingsEditor/MainPage.cpp [hunk 7]
```diff
@@ -295,6 +296,17 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
         {
             contentFrame().Navigate(xaml_typename<Editor::Actions>(), winrt::make<ActionsPageNavigationState>(_settingsClone));
         }
+        else if (clickedItemTag == globalProfileTag)
```
[the remark this answers is not in the document]
Comment (DHowett) [reply src/cascadia/TerminalSettingsEditor/MainPage.cpp:299]: Can't do it. The second part of the condition is not a constant expression (it changes based on a runtime value.)

File: src/cascadia/TerminalSettingsEditor/SettingContainer.cpp (hunk 5 above)
Comment (DHowett) [reply src/cascadia/TerminalSettingsEditor/SettingContainer.cpp:190]:
```suggestion
            const auto profile = appearanceConfig.SourceProfile();
```

File: src/features.xml [hunk 8]
```diff
@@ -49,4 +49,12 @@
             <brandingToken>WindowsInbox</brandingToken>
         </alwaysEnabledBrandingTokens>
     </feature>
+
+    <feature>
+        <name>Feature_ShowProfileDefaultsInSettings</name>
+        <description>Whether to show the "base layer" ("defaults") page in the Terminal settings UI</description>
+        <stage>AlwaysEnabled</stage>
+        <!-- This feature will not ship to Stable until it is complete. -->
```
[the remark this answers is not in the document]
Comment (DHowett) [reply src/features.xml:57]: I feel like we're not gonna forget about this one no matter how we do it

File: src/cascadia/TerminalSettingsEditor/MainPage.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (DHowett) [reply src/cascadia/TerminalSettingsEditor/MainPage.cpp:299]: And also... eh. If somebody uses the xaml inspector to set Visibility to True for this in *Stable* and therefore access the hidden Defaults page? I am _okay_ with that level of weird.

File: src/features.xml (hunk 8 above)
[the remark this answers is not in the document]
Comment (DHowett) [reply src/features.xml:58]: i added 10430 to the `<id>` field in the feature doc

**Review by miniksa — APPROVED** [review]
